components/signup/signup.tsx

```diff
@@ -0,0 +1,692 @@
+                            max: Constants.MAX_USERNAME_LENGTH,
+                        },
+                    )
+                ));
+                isValid = false;
+            }
+        } else {
+            setNameError(formatMessage({id: 'signup_user_completed.required', defaultMessage: 'This field is required'}));
+            isValid = false;
+        }
+
+        const providedPassword = passwordInput.current?.value ?? '';
+        const {error} = isValidPassword(providedPassword, getPasswordConfig(config), intl);
+
+        if (error) {
+            setPasswordError(error as string);
+            isValid = false;
+        }
+
+        return isValid;
+    };
+
+    const dismissAlert = () => {
+        setAlertBanner(null);
+    };
+
+    const handleSubmit = async (e: React.MouseEvent | React.KeyboardEvent) => {
+        e.preventDefault();
+        trackEvent('signup_email', 'click_create_account');
+        setIsWaiting(true);
+
+        if (isUserValid()) {
+            setNameError('');
+            setEmailError('');
+            setPasswordError('');
+            setServerError('');
+            setIsWaiting(true);
+
+            const user = {
+                email: emailInput.current?.value.trim(),
+                username: nameInput.current?.value.trim().toLowerCase(),
+                password: passwordInput.current?.value,
+                allow_marketing: true,
+            } as UserProfile;
+
+            const redirectTo = (new URLSearchParams(search)).get('redirect_to') as string;
+
+            const {data, error} = await dispatch(createUser(user, token, inviteId, redirectTo));
+
+            if (error) {
+                setAlertBanner({
+                    mode: 'danger' as ModeType,
+                    title: (error as ServerError).message,
+                    onDismiss: dismissAlert,
+                });
+                setIsWaiting(false);
+                return;
+            }
+
+            await handleSignupSuccess(user, data as UserProfile);
+        }
+
+        setIsWaiting(false);
+    };
+
+    const handleReturnButtonOnClick = () => history.replace('/');
+
+    const getContent = () => {
+        if (!enableSignUpWithEmail && !enableExternalSignup) {
+            return (
+                <ColumnLayout
+                    title={formatMessage({id: 'login.noMethods.title', defaultMessage: 'This server doesn’t have any sign-in methods enabled'})}
+                    message={formatMessage({id: 'login.noMethods.subtitle', defaultMessage: 'Please contact your System Administrator to resolve this.'})}
+                />
+            );
+        }
+
+        if (noOpenServer || serverError || usedBefore) {
+            return (
+                <ColumnLayout
+                    title={noOpenServer ? (
+                        formatMessage({id: 'signup_user_completed.no_open_server.title', defaultMessage: 'This server doesn’t allow open signups'})
```

Comment:
nit: this logic here is a bit hard to read. Would you mind extracting this in a variable before the return?

Reply:
Updated